File: tools/covenant/lifecycle_events.py
#!/usr/bin/env python3
"""Publish sub-agent lifecycle events to Cortana event bus."""

from __future__ import annotations

import argparse
import json
import os
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

PSQL_BIN = "/opt/homebrew/opt/postgresql@17/bin/psql"
DEFAULT_DB = "cortana"
EVENT_SOURCE = "agent_lifecycle"
TRACE_CLI = Path("/Users/hd/openclaw/tools/covenant/trace.py")


class LifecycleEventError(Exception):
    pass


def sql_quote(value: str) -> str:
    return value.replace("'", "''")


def run_psql(db: str, sql: str) -> str:
    env = os.environ.copy()
    env["PATH"] = "/opt/homebrew/opt/postgresql@17/bin:" + env.get("PATH", "")

    proc = subprocess.run(
        [PSQL_BIN, db, "-X", "-q", "-At", "-c", sql],
        capture_output=True,
        text=True,
        env=env,
    )
    if proc.returncode != 0:
        raise LifecycleEventError(proc.stderr.strip() or "psql command failed")
    return proc.stdout.strip()


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def log_trace_span(
    db: str,
    trace_id: str | None,
    span_name: str,
    agent_role: str,
    task_id: int,
    chain_id: str,
    started_at: str,
    metadata: dict[str, Any] | None = None,
) -> None:
    if not trace_id or not TRACE_CLI.exists():
        return

    meta = metadata or {}
    cmd = [
        "python3",
        str(TRACE_CLI),
        "--db",
        db,
        "log",
        trace_id,
        span_name,
        "--agent",
        agent_role,
        "--task",
        str(task_id),
        "--chain-id",
        chain_id,
        "--start",
        started_at,
        "--end",
        _now_iso(),
        "--metadata",
        json.dumps(meta, ensure_ascii=False),
    ]

    subprocess.run(cmd, capture_output=True, text=True)


def publish_event(db: str, event_type: str, payload: dict[str, Any]) -> int:
    payload_json = json.dumps(payload, ensure_ascii=False)
    sql = (
        "SELECT cortana_event_bus_publish("
        f"'{sql_quote(event_type)}', "
        f"'{EVENT_SOURCE}', "
        f"'{sql_quote(payload_json)}'::jsonb, "
        "NULL"
        ");"
    )
    out = run_psql(db, sql)
    try:
        return int(out)
    except ValueError as exc:
        raise LifecycleEventError(f"Unexpected publish result: {out!r}") from exc


def publish_spawn(
    agent_role: str,
    task_id: int,
    chain_id: str,
    label: str,
    model: str,
    trace_id: str | None = None,
    db: str = DEFAULT_DB,
) -> int:
    return publish_event(
        db,
        "agent_spawned",
        {
            "agent_role": agent_role,
            "task_id": task_id,
            "chain_id": chain_id,
            "trace_id": trace_id,
            "label": label,
            "model": model,
        },
    )


def publish_completion(
    agent_role: str,
    task_id: int,
    chain_id: str,
    label: str,
    duration_ms: int,
    outcome_summary: str,
    trace_id: str | None = None,
    db: str = DEFAULT_DB,
) -> int:
    return publish_event(
        db,
        "agent_completed",
        {
            "agent_role": agent_role,
            "task_id": task_id,
            "chain_id": chain_id,
            "trace_id": trace_id,
            "label": label,
            "duration_ms": duration_ms,
            "outcome_summary": outcome_summary,
        },
    )


def publish_failure(
    agent_role: str,
    task_id: int,
    chain_id: str,
    label: str,
    error: str,
    duration_ms: int,
    trace_id: str | None = None,
    db: str = DEFAULT_DB,
) -> int:
    return publish_event(
        db,
        "agent_failed",
        {
            "agent_role": agent_role,
            "task_id": task_id,
            "chain_id": chain_id,
            "trace_id": trace_id,
            "label": label,
            "error": error,
            "duration_ms": duration_ms,
        },
    )


def publish_timeout(
    agent_role: str,
    task_id: int,
    chain_id: str,
    label: str,
    timeout_seconds: int,
    trace_id: str | None = None,
    db: str = DEFAULT_DB,
) -> int:
    return publish_event(
        db,
        "agent_timeout",
        {
            "agent_role": agent_role,
            "task_id": task_id,
            "chain_id": chain_id,
            "trace_id": trace_id,
            "label": label,
            "timeout_seconds": timeout_seconds,
        },
    )


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Publish agent lifecycle event")
    parser.add_argument("--db", default=DEFAULT_DB, help="PostgreSQL database (default: cortana)")

    sub = parser.add_subparsers(dest="command", required=True)

    base_parent = argparse.ArgumentParser(add_help=False)
    base_parent.add_argument("--agent-role", required=True)
    base_parent.add_argument("--task-id", type=int, required=True)
    base_parent.add_argument("--chain-id", required=True)
    base_parent.add_argument("--trace-id", help="Correlation trace id (UUID)")
    base_parent.add_argument("--label", required=True)

    sp = sub.add_parser("spawn", parents=[base_parent], help="Publish agent_spawned")
    sp.add_argument("--model", required=True)

    cp = sub.add_parser("complete", parents=[base_parent], help="Publish agent_completed")
    cp.add_argument("--duration-ms", type=int, required=True)
    cp.add_argument("--outcome-summary", required=True)

    fp = sub.add_parser("fail", parents=[base_parent], help="Publish agent_failed")
    fp.add_argument("--error", required=True)
    fp.add_argument("--duration-ms", type=int, required=True)

    tp = sub.add_parser("timeout", parents=[base_parent], help="Publish agent_timeout")
    tp.add_argument("--timeout-seconds", type=int, required=True)

    return parser


def main() -> int:
    parser = build_parser()
    args = parser.parse_args()

    try:
        started_at = _now_iso()

        if args.command == "spawn":
            event_id = publish_spawn(
                args.agent_role,
                args.task_id,
                args.chain_id,
                args.label,
                args.model,
                trace_id=args.trace_id,
                db=args.db,
            )
            event_type = "agent_spawned"
            log_trace_span(
                args.db,
                args.trace_id,
                "agent_spawn",
                args.agent_role,
                args.task_id,
                args.chain_id,
                started_at,
                {"label": args.label, "model": args.model},
            )
        elif args.command == "complete":
            event_id = publish_completion(
                args.agent_role,
                args.task_id,
                args.chain_id,
                args.label,
                args.duration_ms,
                args.outcome_summary,
                trace_id=args.trace_id,
                db=args.db,
            )
            event_type = "agent_completed"
            log_trace_span(
                args.db,
                args.trace_id,
                "agent_complete",
                args.agent_role,
                args.task_id,
                args.chain_id,
                started_at,
                {"label": args.label, "outcome_summary": args.outcome_summary},
            )
        elif args.command == "fail":
            event_id = publish_failure(
                args.agent_role,
                args.task_id,
                args.chain_id,
                args.label,
                args.error,
                args.duration_ms,
                trace_id=args.trace_id,
                db=args.db,
            )
            event_type = "agent_failed"
            log_trace_span(
                args.db,
                args.trace_id,
                "agent_fail",
                args.agent_role,
                args.task_id,
                args.chain_id,
                started_at,
                {"label": args.label, "error": args.error},
            )
        elif args.command == "timeout":
            event_id = publish_timeout(
                args.agent_role,
                args.task_id,
                args.chain_id,
                args.label,
                args.timeout_seconds,
                trace_id=args.trace_id,
                db=args.db,
            )
            event_type = "agent_timeout"
            log_trace_span(
                args.db,
                args.trace_id,
                "agent_timeout",
                args.agent_role,
                args.task_id,
                args.chain_id,
                started_at,
                {"label": args.label, "timeout_seconds": args.timeout_seconds},
            )
        else:
            raise LifecycleEventError(f"Unsupported command: {args.command}")

        print(json.dumps({"ok": True, "event_id": event_id, "event_type": event_type, "trace_id": args.trace_id}))
        return 0
    except LifecycleEventError as exc:
        print(f"LIFECYCLE_EVENT_ERROR: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/covenant/lifecycle_events.ts
#!/usr/bin/env npx tsx

/** Publish sub-agent lifecycle events to Cortana event bus. */

import fs from "fs";
import path from "path";
import { spawnSync } from "child_process";
import { runPsql, withPostgresPath } from "../lib/db.js";
import { resolveRepoPath } from "../lib/paths.js";

const DEFAULT_DB = "cortana";
const EVENT_SOURCE = "agent_lifecycle";
const TRACE_CLI = path.join(resolveRepoPath(), "tools", "covenant", "trace.ts");

class LifecycleEventError extends Error {}

type Json = Record<string, any>;

type BaseArgs = {
  agentRole: string;
  taskId: number;
  chainId: string;
  traceId?: string;
  label: string;
};

function sqlQuote(value: string): string {
  return value.replace(/'/g, "''");
}

function runPsqlQuery(db: string, sql: string): string {
  const result = runPsql(sql, { db, args: ["-X", "-q", "-At"], env: withPostgresPath(process.env) });
  if (result.status !== 0) {
    const err = (result.stderr || "").toString().trim();
    throw new LifecycleEventError(err || "psql command failed");
  }
  return (result.stdout || "").toString().trim();
}

function nowIso(): string {
  return new Date().toISOString().replace("Z", "+00:00");
}

function logTraceSpan(
  db: string,
  traceId: string | undefined,
  spanName: string,
  agentRole: string,
  taskId: number,
  chainId: string,
  startedAt: string,
  metadata?: Json
): void {
  if (!traceId || !fs.existsSync(TRACE_CLI)) return;

  const meta = metadata ?? {};
  const cmd = [
    TRACE_CLI,
    "--db",
    db,
    "log",
    traceId,
    spanName,
    "--agent",
    agentRole,
    "--task",
    String(taskId),
    "--chain-id",
    chainId,
    "--start",
    startedAt,
    "--end",
    nowIso(),
    "--metadata",
    JSON.stringify(meta),
  ];

  spawnSync(cmd[0], cmd.slice(1), { encoding: "utf8" });
}

function publishEvent(db: string, eventType: string, payload: Json): number {
  const payloadJson = JSON.stringify(payload);
  const sql =
    "SELECT cortana_event_bus_publish(" +
    `'${sqlQuote(eventType)}', ` +
    `'${EVENT_SOURCE}', ` +
    `'${sqlQuote(payloadJson)}'::jsonb, ` +
    "NULL" +
    ");";
  const out = runPsqlQuery(db, sql);
  const parsed = Number.parseInt(out, 10);
  if (!Number.isFinite(parsed)) {
    throw new LifecycleEventError(`Unexpected publish result: ${JSON.stringify(out)}`);
  }
  return parsed;
}

function publishSpawn(
  agentRole: string,
  taskId: number,
  chainId: string,
  label: string,
  model: string,
  traceId: string | undefined,
  db: string
): number {
  return publishEvent(db, "agent_spawned", {
    agent_role: agentRole,
    task_id: taskId,
    chain_id: chainId,
    trace_id: traceId,
    label,
    model,
  });
}

function publishCompletion(
  agentRole: string,
  taskId: number,
  chainId: string,
  label: string,
  durationMs: number,
  outcomeSummary: string,
  traceId: string | undefined,
  db: string
): number {
  return publishEvent(db, "agent_completed", {
    agent_role: agentRole,
    task_id: taskId,
    chain_id: chainId,
    trace_id: traceId,
    label,
    duration_ms: durationMs,
    outcome_summary: outcomeSummary,
  });
}

function publishFailure(
  agentRole: string,
  taskId: number,
  chainId: string,
  label: string,
  error: string,
  durationMs: number,
  traceId: string | undefined,
  db: string
): number {
  return publishEvent(db, "agent_failed", {
    agent_role: agentRole,
    task_id: taskId,
    chain_id: chainId,
    trace_id: traceId,
    label,
    error,
    duration_ms: durationMs,
  });
}

function publishTimeout(
  agentRole: string,
  taskId: number,
  chainId: string,
  label: string,
  timeoutSeconds: number,
  traceId: string | undefined,
  db: string
): number {
  return publishEvent(db, "agent_timeout", {
    agent_role: agentRole,
    task_id: taskId,
    chain_id: chainId,
    trace_id: traceId,
    label,
    timeout_seconds: timeoutSeconds,
  });
}

function usageError(): never {
  console.error("usage: lifecycle_events.py [--db <db>] {spawn|complete|fail|timeout} <args>");
  process.exit(2);
}

function parseBaseArgs(args: string[]): BaseArgs {
  const get = (flag: string): string | undefined => {
    const idx = args.indexOf(flag);
    if (idx >= 0) return args[idx + 1];
    const eq = args.find((a) => a.startsWith(`${flag}=`));
    if (eq) return eq.slice(flag.length + 1);
    return undefined;
  };

  const agentRole = get("--agent-role");
  const taskIdRaw = get("--task-id");
  const chainId = get("--chain-id");
  const label = get("--label");
  const traceId = get("--trace-id");

  if (!agentRole || !taskIdRaw || !chainId || !label) usageError();

  return {
    agentRole,
    taskId: Number(taskIdRaw),
    chainId,
    traceId,
    label,
  };
}

async function main(): Promise<void> {
  const argv = process.argv.slice(2);
  let db = DEFAULT_DB;
  for (let i = 0; i < argv.length; i += 1) {
    if (argv[i] === "--db" && argv[i + 1]) {
      db = argv[i + 1];
      argv.splice(i, 2);
      i -= 1;
      continue;
    }
    if (argv[i].startsWith("--db=")) {
      db = argv[i].slice("--db=".length);
      argv.splice(i, 1);
      i -= 1;
    }
  }

  const command = argv.shift();
  if (!command) usageError();

  try {
    const startedAt = nowIso();

    if (command === "spawn") {
      const base = parseBaseArgs(argv);
      const model = (() => {
        const idx = argv.indexOf("--model");
        if (idx >= 0) return argv[idx + 1];
        const eq = argv.find((a) => a.startsWith("--model="));
        if (eq) return eq.slice("--model=".length);
        return undefined;
      })();
      if (!model) usageError();

      const eventId = publishSpawn(base.agentRole, base.taskId, base.chainId, base.label, model, base.traceId, db);
      logTraceSpan(db, base.traceId, "agent_spawn", base.agentRole, base.taskId, base.chainId, startedAt, {
        label: base.label,
        model,
      });
      console.log(JSON.stringify({ ok: true, event_id: eventId, event_type: "agent_spawned", trace_id: base.traceId }));
      return;
    }

    if (command === "complete") {
      const base = parseBaseArgs(argv);
      const durationMsRaw = (() => {
        const idx = argv.indexOf("--duration-ms");
        if (idx >= 0) return argv[idx + 1];
        const eq = argv.find((a) => a.startsWith("--duration-ms="));
        if (eq) return eq.slice("--duration-ms=".length);
        return undefined;
      })();
      const outcomeSummary = (() => {
        const idx = argv.indexOf("--outcome-summary");
        if (idx >= 0) return argv[idx + 1];
        const eq = argv.find((a) => a.startsWith("--outcome-summary="));
        if (eq) return eq.slice("--outcome-summary=".length);
        return undefined;
      })();
      if (!durationMsRaw || !outcomeSummary) usageError();
      const durationMs = Number(durationMsRaw);

      const eventId = publishCompletion(
        base.agentRole,
        base.taskId,
        base.chainId,
        base.label,
        durationMs,
        outcomeSummary,
        base.traceId,
        db
      );
      logTraceSpan(db, base.traceId, "agent_complete", base.agentRole, base.taskId, base.chainId, startedAt, {
        label: base.label,
        outcome_summary: outcomeSummary,
      });
      console.log(JSON.stringify({ ok: true, event_id: eventId, event_type: "agent_completed", trace_id: base.traceId }));
      return;
    }

    if (command === "fail") {
      const base = parseBaseArgs(argv);
      const errorText = (() => {
        const idx = argv.indexOf("--error");
        if (idx >= 0) return argv[idx + 1];
        const eq = argv.find((a) => a.startsWith("--error="));
        if (eq) return eq.slice("--error=".length);
        return undefined;
      })();
      const durationMsRaw = (() => {
        const idx = argv.indexOf("--duration-ms");
        if (idx >= 0) return argv[idx + 1];
        const eq = argv.find((a) => a.startsWith("--duration-ms="));
        if (eq) return eq.slice("--duration-ms=".length);
        return undefined;
      })();
      if (!errorText || !durationMsRaw) usageError();
      const durationMs = Number(durationMsRaw);

      const eventId = publishFailure(
        base.agentRole,
        base.taskId,
        base.chainId,
        base.label,
        errorText,
        durationMs,
        base.traceId,
        db
      );
      logTraceSpan(db, base.traceId, "agent_fail", base.agentRole, base.taskId, base.chainId, startedAt, {
        label: base.label,
        error: errorText,
      });
      console.log(JSON.stringify({ ok: true, event_id: eventId, event_type: "agent_failed", trace_id: base.traceId }));
      return;
    }

    if (command === "timeout") {
      const base = parseBaseArgs(argv);
      const timeoutRaw = (() => {
        const idx = argv.indexOf("--timeout-seconds");
        if (idx >= 0) return argv[idx + 1];
        const eq = argv.find((a) => a.startsWith("--timeout-seconds="));
        if (eq) return eq.slice("--timeout-seconds=".length);
        return undefined;
      })();
      if (!timeoutRaw) usageError();
      const timeoutSeconds = Number(timeoutRaw);

      const eventId = publishTimeout(
        base.agentRole,
        base.taskId,
        base.chainId,
        base.label,
        timeoutSeconds,
        base.traceId,
        db
      );
      logTraceSpan(db, base.traceId, "agent_timeout", base.agentRole, base.taskId, base.chainId, startedAt, {
        label: base.label,
        timeout_seconds: timeoutSeconds,
      });
      console.log(JSON.stringify({ ok: true, event_id: eventId, event_type: "agent_timeout", trace_id: base.traceId }));
      return;
    }

    throw new LifecycleEventError(`Unsupported command: ${command}`);
  } catch (err) {
    if (err instanceof LifecycleEventError) {
      console.error(`LIFECYCLE_EVENT_ERROR: ${err.message}`);
      process.exit(1);
    }
    throw err;
  }
}

main().catch((err) => {
  console.error(err instanceof Error ? err.stack || err.message : String(err));
  process.exit(1);
});
